src/allocation_engine/interval_tree.rs

```diff
@@ -432,85 +332,364 @@ impl InnerNode {
 
-/// Compute height of the optional sub-tree.
-#[allow(dead_code)]
-fn height(node: &Option<Box<InnerNode>>) -> u64 {
-    node.as_ref().map_or(0, |n| n.height)
+/// An interval tree implementation specialized for VMM memory slots management.
+#[derive(Clone, Debug, PartialEq, PartialOrd, Eq, Ord)]
+pub struct IntervalTree {
+    root: Option<Node>,
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
+impl IntervalTree {
+    /// Creates a new IntervalTree object that is going to be used by the
+    /// AddressAllocator.
+    #[allow(dead_code)]
+    pub fn new(key: Range, node_state: NodeState) -> Self {
+        IntervalTree {
+            root: Some(Node::new(key, node_state)),
+        }
+    }
 
-    #[test]
-    fn test_new_range() {
-        assert_eq!(Range::new(2, 1).unwrap_err(), Error::InvalidRange(2, 1));
+    #[allow(dead_code)]
+    fn search(&self, key: &Range) -> Option<&Node> {
+        match self.root {
+            None => None,
+            Some(ref node) => node.search(key),
+        }
     }
 
-    #[test]
-    fn test_range_overlaps() {
-        let range_a = Range::new(1u64, 4u64).unwrap();
-        let range_b = Range::new(4u64, 6u64).unwrap();
-        let range_c = Range::new(2u64, 3u64).unwrap();
-        let range_d = Range::new(4u64, 4u64).unwrap();
-        let range_e = Range::new(5u64, 6u64).unwrap();
-
-        assert!(range_a.overlaps(&range_b));
-        assert!(range_b.overlaps(&range_a));
-        assert!(range_a.overlaps(&range_c));
-        assert!(range_c.overlaps(&range_a));
-        assert!(range_a.overlaps(&range_d));
-        assert!(range_d.overlaps(&range_a));
-        assert!(!range_a.overlaps(&range_e));
-        assert!(!range_e.overlaps(&range_a));
-
-        assert_eq!(range_a.len(), 4);
-        assert_eq!(range_d.len(), 1);
+    #[allow(dead_code)]
+    fn search_superset(&self, key: &Range) -> Option<&Node> {
+        match self.root {
+            None => None,
+            Some(ref node) => node.search_superset(key),
+        }
     }
 
-    #[test]
-    fn test_range_contain() {
-        let range_a = Range::new(2u64, 6u64).unwrap();
-        assert!(range_a.contains(&Range::new(2u64, 3u64).unwrap()));
-        assert!(range_a.contains(&Range::new(3u64, 4u64).unwrap()));
-        assert!(range_a.contains(&Range::new(5u64, 5u64).unwrap()));
-        assert!(range_a.contains(&Range::new(5u64, 6u64).unwrap()));
-        assert!(range_a.contains(&Range::new(6u64, 6u64).unwrap()));
-        assert!(!range_a.contains(&Range::new(1u64, 1u64).unwrap()));
-        assert!(!range_a.contains(&Range::new(1u64, 2u64).unwrap()));
-        assert!(!range_a.contains(&Range::new(1u64, 3u64).unwrap()));
-        assert!(!range_a.contains(&Range::new(1u64, 7u64).unwrap()));
-        assert!(!range_a.contains(&Range::new(7u64, 8u64).unwrap()));
-        assert!(!range_a.contains(&Range::new(6u64, 7u64).unwrap()));
-        assert!(!range_a.contains(&Range::new(7u64, 8u64).unwrap()));
+    #[allow(dead_code)]
```

Comment:
Why do we still need to allow dead code?

Reply:
Because I have not added examples. Will create a new PR after this one to add examples, update documentation and READMEs and will remove the dead code then. Is this okay?

Reply:
The examples should not influence the `allow(dead_code)` macro. If there is still a warning from clippy regarding dead code, it means that we are not exporting things properly. Since from the implementation stand point this is complete, it would be better to remove `allow(dead_code)` now to make sure that the usage is not the one from the examples.